### PR TITLE
implemented MutableGenotype

### PR DESCRIPTION
**Pull request by jbloom22**
- added test to GenotypeStreamSuite
- moved Boolean functions like isHet back to Genotype object as functions on gt: Int, for wider use with hardCallIterator

Eager for feedback on these steps and next steps.

### CODE REVIEW

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 1]
```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
...
 
+class MutableGenotype(nAlleles: Int) extends Genotype {
+
+  private var _gt: Int = 0
+  private val _ad: Array[Int] = Array.ofDim[Int](nAlleles)
+  private var _dp: Int = 0
+  private var _gq: Int = 0
+  private val _px: Array[Int] = Array.ofDim[Int](triangle(nAlleles))
+  var _fakeRef: Boolean = false
+  var _isDosage: Boolean = false
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: Should these be private?

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 2]
```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
...
+
+  if (isDosage) {
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: This is never run.  Dead code.

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 3]
```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
...
+      require(_px.sum == 32768)
+      require(_gt == Genotype.gtFromLinear(_px).getOrElse(-1))
+    }
+  }
+
+  def unboxedGT: Int = _gt
+  def fakeRef: Boolean = _fakeRef
+  def isDosage: Boolean = _isDosage
+
+  def ad: Option[Array[Int]] = if (_ad(0) >= 0) Some(_ad) else None
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: I wouldn't reuse `_ad(0)` this way.  Just make a `private var hasAD: Boolean = false`.

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 4]
```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
...
+
+  def dosage: Option[Array[Double]] =
+    if (_px(0) >= 0)
+      None
+    else if (isDosage)
+      Some(_px.map(_ / 32768.0))
+    else
+      Some(Genotype.phredToDosage(_px))
+
+  def set(nAlleles: Int, isDosage: Boolean, a: ByteIterator) {
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: `read` would be more consistent with the interface elsewhere.

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 5]
```diff
@@ -120,51 +120,25 @@ abstract class Genotype extends Serializable {
 
   def dosage: Option[Array[Double]]
 
-  def isHomRef: Boolean = unboxedGT == 0
+  def isHomRef: Boolean = Genotype.isHomRef(unboxedGT)
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: Why was this necessary?

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 6]
```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
       Some(Genotype.phredToDosage(_px))
 }
 
+class MutableGenotype(nAlleles: Int) extends Genotype {
+
+  private var _gt: Int = 0
```
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: These should probably default to missing, all of them.

**Comment by cseed** [comment]
Looks good!  Next steps:

 - Use it!
 - Test it!  I'm not sure how much will break.
 - Time it!  Do something simple filter genotypes gq >= 20 and a sampleqc or something.  Does it help?  How much?
 - Figure out how you're going to deal with annotations like `va.rare_genos = gs.filter(g => ... some rare condition ...).collect()`.

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 5 above)
Comment (jbloom22) [inline src/main/scala/is/hail/variant/Genotype.scala]: Moved in preparation for Issue #1314.

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 1 above)
Comment (jbloom22) [inline src/main/scala/is/hail/variant/Genotype.scala]: Yes

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 3 above)
Comment (jbloom22) [inline src/main/scala/is/hail/variant/Genotype.scala]: Done

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 2 above)
Comment (jbloom22) [inline src/main/scala/is/hail/variant/Genotype.scala]: Killed

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 4 above)
Comment (jbloom22) [inline src/main/scala/is/hail/variant/Genotype.scala]: Agreed
